man/man1/cmpbin.1
.TH cmpbin 1  "USER COMMANDS"
.SH NAME
cmpbin \- compares two binary output files
.SH SYNOPSIS
.B cmpbin
[\-metafile] [\-verbose] [\-csv] [\-short] \fIfile1.bin\fP \fIfile2.bin\fP

.SH DESCRIPTION
Program compares two binary output files. The maximum difference and its location, the sum of all absolute differences and the average difference are printed.
.SH OPTIONS
.TP
\-metafile
Output files are JSON metafiles describing the structure of the raw binary file.
.TP
\-verbose
Print values and absolute difference for each different data item.
.TP
\-csv
Print output in comma-separated format.
.TP
\-short
Datatype of binary data is short, default is float. If \-metafile option is set then datatype is read from JSON metafile.
.TP
.I file1.bin file2.bin
Filenames of binary output data files to compare.
.SH EXAMPLE
.TP
Compare two binary files in the current directory:
.B cmpbin
vegc1.bin vegc2.bin
.TP
or
.B cmpbin
-metafile vegc1.bin.json vegc2.bin.json
.PP
.SH EXIT STATUS
.B cmpbin
returns a zero exit status if binary files could be compared.
Non zero is returned in case of failure.

.SH AUTHORS

For authors and contributors see AUTHORS file

.SH COPYRIGHT

(C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file

.SH SEE ALSO
bin2cdf(1), cdf2bin(1), lpjml(1)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project follows a [GNU standard version numbering](https://www.gnu.org/prep/standards/html_node/Releases.html#index-version-numbers_002c-for-releases)
of `major.minor.patch` with
- `major` for substantial changes to code functionality,
- `minor` for additions of modules and functionality, and
- `patch` for technical changes and bugfixes without extending the functionality of the code.

- if applicable, categories to include in the CHANGELOG.md are:
  - Added
  - Changed
  - Deprecated
  - Removed
  - Fixed
  - Security

## [Unreleased]


### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de)
- code review:

### Added

- Check for valid `swc_bnf` parameters added in `fscanpftpar.c`.
- Wirth et al. (2024)  article added to `REFERENCES`.

### Changed

- `phi_bnf` parameter calculated from `swc_bnf` parameter instead of reading from `pft.cjson` file in `fscanpftpar.c`.

### Removed

- Parameter `phi_bnf` removed from `pft.cjson`.


## [5.10.0] - 2025-08-14

### Contributors

- author: Jens Heinke (heinke@pik-potsdam.de), Sibyll Schaphoff (sibylls@pik-potsdam.de), Stephen Wirth (wirth@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de)
- code review: Sibyll Schaphoff (sibylls@pik-potsdam.de), Sebastian Ostberg (ostberg@pik-potsdam.de)

### Added

- option to specify year after which deforestation is no longer feeding product pools but biomass is burnt instead: `luc_burn_startyear` (disabled)
- option to specify costs for N recovery upon turnover (pre-compiler switch `NRECOVERY_COST`, disabled)

### Changed

- parameter changes: 
  - `alphaa` from 0.5 to 0.7 for natural vegetation; 
  - `theta` from 0.9 to 0.95; 
  - `fn_turnover` PFT1: 0.5, PFT2: 0.5, PFT3: 0.7, PFT4: 0.3, PFT5: 0.3, PFT6: 0.7, PFT7: 0.3, PFT8: 0.3; updated values from [Sophia et al. 2024](https://doi.org/10.5194/bg-21-4169-2024) for all other PFTs except crops
  - `cnratio_leaf.low/median/high` for all PFTs updated from TRY data
  - `b` (ratio of dark respiration Rd to maximum photosynthetic capacity Vcmax) set to 0.031 for C3 and C4 plants (value for 25°C, from [Wang et al. 2020](https://doi.org/10.1111/gcb.14980), consistent with TRY data)
  - `aphen_min` from 60 to 90 and `aphen_max` from 245 to 180
  - `nfixpot` from 0.01 to 0.5 (middle of proposed range in [Yu and Zhuang 2020](https://doi.org/10.5194/bg-17-3643-2020))
  - `turnover` parameter for leaves&roots set to 2 for TeNE and 3 BoNE 
  - temperature boundary between temperate and boreal zone changed from -2°C to -5°C
- acclimation of `b` to mean vegetation period temperature (following [Wang et al. 2020](https://doi.org/10.1111/gcb.14980))
- calculation of structural leaf N content (leaf N independent of Vcmax) based on `ncleaf.low` in both `ndemand_xx` and `vmaxlimit_xx` (was `ncleaf.medium` in `ndemand_xx`)
- scaling of vmax with f_LAI removed
- N recovery from turnover directly added to `bm_inc.nitrogen`
- N uptake now separated for uptake of NH4 and NO3 with parameters from  [Craig et al. 2025](https://doi.org/10.1111/nph.70140) (median of values)
- trees are burnt (instead of added to the litter) when running with `luc_timber=FALSE`
- beginning of vegetative period for tropical raingreen trees set at beginning of 6 month period with highest precipitation sum
- leaf shedding for deciduous trees induced after `aphen_max` or when `phen<0.1` after `aphen_min`
- `phen` permanently set to 1 for tropical evergreen trees

### Removed

### Fixed

- corrected conversion of vmax in computation of N demand, which was incorrectly scaled with `24/daylength`
- corrected temperature sensitivity of N demand (`k_temp` changed from 0.02 to 0.0693)
- corrected `f_NCplant` to comply with equation from [Smith et al. 2014](https://doi.org/10.5194/bg-11-2027-2014)
- corrected phenology of trees:
  - leaf turnover of raingreen trees at leaf shedding (same as for summergreen trees, had continuous daily leaf turnover like evergreen trees)
  - `phen` set to zero at leaf shedding until restart of phenological cycle to prevent phen>1 without leaf carbon 


## [5.9.28] - 2025-07-31

### Contributors

- author: Werner von Bloh (bloh@pik-potsdam.de)
- code review: Maik Billing (billing@pik-potsdam.de), Christoph Müller (cmueller@pik-potsdam.de), Sebastian Ostberg(ostberg@pik-potsdam.de), Jens Heinke (heinke@pik-potsdam.de)

### Added

- Option `-h` added to utilities `addheader`, `cdf2bin`, `regridclm`, `regridsoil` and `printclm` to print help text.
- Notice added in `configure.sh` that `make clean; make all` has to be performed after change in configuration.
- Support for short datatype added in `cmpbin` utility.
- Shell script `allbin2cdf` added to convert all binary raw output files in a directory to NetCDF files.
- Option `-nounit` added to `bin2cdf` to set unit if unit is defined in the metafile as an empty string.

### Changed

- If option `-ncell 0` is used in `addheader` utility then the number of cells is calculated from the file size of the binary file.
- Utilities `grid2clm` and `cft2clm` are replaced by alias to `addheader`.
- File `default.md` converted from DOS to Linux format.
- If scaling factor is set as an option in `bin2cdf` then this scaling factor is used insteads that one defined in the JSON metafile.
- Man pages updated.

### Fixed

- Missing `free(axis->comment)` added in `freeaxis()` to avoid memory leak.
- Missing deallocation of list added in `open_config.c`.
- Function `nc_close()` replaced by `closeclimate_netcdf()` in `cdf2clm.c` and `cdf2bin.c` to avoid memory leak.
- Argument for error message corrected in `getcellindex.c` and `getcountry.c`.
- Typo in man page of `configure.sh` fixed.
- Comment for `pft` initialized in `initsetting_netcdf.c`.
- Access to undefined `pft_name.comment` removed in `bin2cdf.c` to avoid SEGV.

src/utils/cmpbin.c
/**************************************************************************************/
/**                                                                                \n**/
/**                       c  m  p  b  i  n  .  c                                   \n**/
/**                                                                                \n**/
/**     Program compares two binary output files                                   \n**/
/**                                                                                \n**/
/** (C) Potsdam Institute for Climate Impact Research (PIK), see COPYRIGHT file    \n**/
/** authors, and contributors see AUTHORS file                                     \n**/
/** This file is part of LPJmL and licensed under GNU AGPL Version 3               \n**/
/** or later. See LICENSE file or go to http://www.gnu.org/licenses/               \n**/
/** Contact: https://github.com/PIK-LPJmL/LPJmL                                    \n**/
/**                                                                                \n**/
/**************************************************************************************/

#include "lpj.h"

#define USAGE "Usage: %s [-metafile] [-verbose] [-csv] [-short] file1.bin file2.bin\n"

int main(int argc,char **argv)
{
  long n,nmax,count;
  size_t offset;
  long long size1,size2;
  FILE *file1,*file2;
  float val1,val2;
  double diff,max;
  Header header1,header2;
  Bool ismeta,swap1,swap2,verbose,iscsv;
  int iarg;
  short sval1,sval2;
  ismeta=swap1=swap2=verbose=iscsv=FALSE;
  header1.datatype=LPJ_FLOAT;
  header2.datatype=LPJ_FLOAT;
  for(iarg=1;iarg<argc;iarg++)
    if(argv[iarg][0]=='-')
    {
      if(!strcmp(argv[iarg],"-metafile"))
        ismeta=TRUE;
      else if(!strcmp(argv[iarg],"-verbose"))
        verbose=TRUE;
      else if(!strcmp(argv[iarg],"-csv"))
        iscsv=TRUE;
      else if(!strcmp(argv[iarg],"-short"))
        header1.datatype=header2.datatype=LPJ_SHORT;
      else
      {
        fprintf(stderr,"Invalid option '%s'.\n",argv[iarg]);
        fprintf(stderr,USAGE,argv[0]);
        return EXIT_FAILURE;
      }
    }
    else
      break;
  if(argc<iarg+2)
  {
    fprintf(stderr,"Missing arguments.\n"
            USAGE,argv[0]);
    return EXIT_FAILURE;
  }
  if(ismeta)
  {
    header1.scalar=1;
    header1.cellsize_lon=header1.cellsize_lat=0.5;
    header1.firstyear=1901;
    header1.firstcell=0;
    header1.nyear=1;
    header1.nbands=1;
    header1.nstep=1;
    header1.timestep=1;
    header1.order=CELLSEQ;
    header1.scalar=1;
    file1=openmetafile(&header1,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,&swap1,&offset,argv[iarg],TRUE);
    if(file1==NULL)
      return EXIT_FAILURE;
    if(fseek(file1,offset,SEEK_CUR))
    {
      fclose(file1);
      fprintf(stderr,"Error seeking in '%s' to offset %lu.\n",argv[iarg],offset);
      return EXIT_FAILURE;
    }
    if(header1.datatype!=LPJ_FLOAT && header1.datatype!=LPJ_SHORT)
    {
      fprintf(stderr,"Datatype in '%s' is %s, must be float or short .\n",argv[iarg],typenames[header1.datatype]);
      fclose(file1);
      return EXIT_FAILURE;
    }
    header2.scalar=1;
    header2.cellsize_lon=header2.cellsize_lat=0.5;
    header2.firstyear=1901;
    header2.firstcell=0;
    header2.nyear=1;
    header2.nbands=1;
    header2.nstep=1;
    header2.timestep=1;
    header2.order=CELLSEQ;
    header2.scalar=1;
    file2=openmetafile(&header2,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,NULL,&swap2,&offset,argv[iarg+1],TRUE);
    if(file2==NULL)
    {
      fclose(file1);
      return EXIT_FAILURE;
    }
    if(fseek(file2,offset,SEEK_CUR))
    {
      fprintf(stderr,"Error seeking in '%s' to offset %lu.\n",argv[iarg+1],offset);
      fclose(file1);
      fclose(file2);
      return EXIT_FAILURE;
    }
    if(header1.nyear!=header2.nyear)
    {
      fprintf(stderr,"Number of years in '%s'=%d differs from number of years in '%s'=%d.\n",
              argv[iarg],header1.nyear,argv[iarg+1],header2.nyear);
      header1.nyear=min(header1.nyear,header2.nyear);
    }
    if(header1.ncell!=header2.ncell)
    {
      fprintf(stderr,"Number of cells in '%s'=%d differs from number of cells in '%s'=%d.\n",
              argv[iarg],header1.ncell,argv[iarg+1],header2.ncell);
      fclose(file1);
      fclose(file2);
      return EXIT_FAILURE;
    }
    if(header1.nbands!=header2.nbands)
    {
      fprintf(stderr,"Number of bands in '%s'=%d differs from number of bands in '%s'=%d.\n",
              argv[iarg],header1.nbands,argv[iarg+1],header2.nbands);
      fclose(file1);
      fclose(file2);
      return EXIT_FAILURE;
    }
    if(header1.nstep!=header2.nstep)
    {
      fprintf(stderr,"Number of steps in '%s'=%d differs from number of steps in '%s'=%d.\n",
              argv[iarg],header1.nstep,argv[iarg+1],header2.nstep);
      fclose(file1);
      fclose(file2);
      return EXIT_FAILURE;
    }
    if(header2.datatype!=header1.datatype)
    {
      fprintf(stderr,"Datatype in '%s' is %s, must be %s of '%s'.\n",
              argv[iarg+1],typenames[header2.datatype],typenames[header1.datatype],argv[iarg]);
      fclose(file1);
      fclose(file2);
      return EXIT_FAILURE;
    }
    count=(long)header1.nyear*header1.ncell*header1.nbands*header1.nstep/header1.timestep;
  }
  else
  {
    file1=fopen(argv[iarg],"rb");
    if(file1==NULL)
    {
      fprintf(stderr,"Error opening '%s': %s\n",argv[iarg],strerror(errno));
      return EXIT_FAILURE;
    }
    file2=fopen(argv[iarg+1],"rb");
    if(file2==NULL)
    {
      fclose(file1);
      fprintf(stderr,"Error opening '%s': %s\n",argv[iarg+1],strerror(errno));
      return EXIT_FAILURE;
    }
    size1=getfilesizep(file1);
    if(size1==0)
    {
      fclose(file1);
      fclose(file2);
      fprintf(stderr,"File '%s' is empty.\n",argv[iarg]);
      return EXIT_FAILURE;
    }
    if(size1 % typesizes[header1.datatype])
    {
      fclose(file1);
      fclose(file2);
      fprintf(stderr,"File size of '%s' is not multiple of %s size.\n",argv[iarg],typenames[header1.datatype]);
      return EXIT_FAILURE;
    }
    size2=getfilesizep(file2);
    if(size2==0)
    {
      fclose(file1);
      fclose(file2);
      fprintf(stderr,"File '%s' is empty.\n",argv[iarg+1]);
      return EXIT_FAILURE;
    }
    if(size2 % typesizes[header2.datatype])
    {
      fclose(file1);
      fclose(file2);
      fprintf(stderr,"File size of '%s' is not multiple of %s size.\n",argv[iarg+1],typenames[header2.datatype]);
      return EXIT_FAILURE;
    }
    if(size1!=size2)
    {
      fprintf(stderr,"File size of '%s'=%lld differs from file size of '%s'=%lld.\n",argv[iarg],size1,argv[iarg+1],size2);
      size1=min(size1,size2);
    }
    count=size1/typesizes[header1.datatype];
  }
  max=0;
  nmax=0;
  diff=0;
  for(n=0;n<count;n++)
  {
    if(header1.datatype==LPJ_FLOAT)
    {
      freadfloat(&val1,1,swap1,file1);
      freadfloat(&val2,1,swap2,file2);
    }
    else
    {
      freadshort(&sval1,1,swap1,file1);
      freadshort(&sval2,1,swap2,file2);
      val1=sval1;
      val2=sval2;
    }
    if(max<fabs(val1-val2))
    {
      max=fabs(val1-val2);
      nmax=n;
    }
    if(verbose && val1!=val2)
    {
      if(ismeta)
        printf("year=%ld,nstep=%ld,nbands=%ld, cell=%ld, ",
               n/header1.ncell/header1.nstep/header1.nbands+header1.firstyear,
               (n/header1.ncell/header1.nbands) % header1.nstep,
               (n/header1.ncell) % header1.nbands,
               n % header1.ncell);
      else
        printf("cell %ld, ",n);
      printf("val1=%g, val2=%g, diff=%g\n",val1,val2,fabs(val1-val2));
    }
    diff+=fabs(val1-val2);
  }
  fclose(file1);
  fclose(file2);
  if(max==0)
  {
    if(iscsv)
      printf((ismeta) ? "0,0,0,0,0,0,0\n" : "0,0,0,0\n");
    else
      printf("Files are identical.\n");
  }
  else
  {
    if(ismeta)
      printf((iscsv) ? "%g,%ld,%ld,%ld,%ld,%g,%g\n" : "max=%g at year=%ld,nstep=%ld,nbands=%ld, cell=%ld, sum=%g,avg=%g\n",
             max,
             nmax/header1.ncell/header1.nstep/header1.nbands+header1.firstyear,
             (nmax/header1.ncell/header1.nbands) % header1.nstep,
             (nmax/header1.ncell) % header1.nbands,
             nmax % header1.ncell,
             diff,diff/n);
    else
      printf((iscsv) ? "%g,%ld,%g,%g\n" : "max=%g at %ld, sum=%g,avg=%g\n",max,nmax,diff,diff/n);
  }
  return EXIT_SUCCESS;
} /* of 'main' */
